src/Components/Components/src/EventCallbackFactoryBinderExtensions.cs

```diff
@@ -11,74 +11,263 @@ namespace Microsoft.AspNetCore.Components
     /// </summary>
     public static class EventCallbackFactoryBinderExtensions
     {
+        private delegate bool BindConverter<T>(object obj, out T value);
+
         // Perf: conversion delegates are written as static funcs so we can prevent
         // allocations for these simple cases.
-        private static Func<object, string> ConvertToString = (obj) => (string)obj;
+        private readonly static BindConverter<string> ConvertToString = ConvertToStringCore;
 
-        private static Func<object, bool> ConvertToBool = (obj) => (bool)obj;
-        private static Func<object, bool?> ConvertToNullableBool = (obj) => (bool?)obj;
+        private static bool ConvertToStringCore(object obj, out string value)
+        {
+            // We expect the input to already be a string.
+            value = (string)obj;
+            return true;
+        }
 
-        private static Func<object, int> ConvertToInt = (obj) => int.Parse((string)obj);
-        private static Func<object, int?> ConvertToNullableInt = (obj) =>
+        private static BindConverter<bool> ConvertToBool = ConvertToBoolCore;
+        private static BindConverter<bool?> ConvertToNullableBool = ConvertToNullableBoolCore;
+
+        private static bool ConvertToBoolCore(object obj, out bool value)
         {
-            if (int.TryParse((string)obj, out var value))
+            // We expect the input to already be a bool.
+            value = (bool)obj;
+            return true;
+        }
+
+        private static bool ConvertToNullableBoolCore(object obj, out bool? value)
+        {
+            // We expect the input to already be a bool.
+            value = (bool?)obj;
+            return true;
+        }
+
+        private static BindConverter<int> ConvertToInt = ConvertToIntCore;
+        private static BindConverter<int?> ConvertToNullableInt = ConvertToNullableIntCore;
+
+        private static bool ConvertToIntCore(object obj, out int value)
+        {
+            var text = (string)obj;
+            if (string.IsNullOrEmpty(text))
             {
-                return value;
+                value = default;
+                return false;
+            }
+
+            if (!int.TryParse(text, out var converted))
+            {
+                value = default;
+                return false;
             }
 
-            return null;
-        };
+            value = converted;
+            return true;
```

Comment:
The difference between nullable and non-nullable:
- Nullable treats empty string as `null` and will bind successfully
- non-nullable fails to parse empty string, and will retain the previous value.

It's problematic to convert an empty value to `default` for non-nullable. It will check the value of the text field to `0` instead of `""`.